upsilonconf/config.py

```diff
@@ -404,17 +404,56 @@ def _replace_in_keys(
     config : Mapping
         The configuration object (or dictionary) to be modified.
     key_modifiers : dict
-        The dictionary with the replacements.
+        The dictionary with the replacements: All keys strings are replaced
+        with the corresponding values from the `key_modifiers` dictionary.
 
     Returns
     -------
     dict
         A dictionary with the modified keys.
 
+    Replacement strategy
+    --------------------
+    1. The replacement should be zealous, meaning that
+       `key_modifiers={'  ': '_', ' ': '-'}` should replace a single spaces
+       with a hyphen and two spaces with an underscore (and not two hypens).
+    2. The replacement should not be recursive, meaning that
+       `key_modifiers={' ': '-', '-': '_'}` should replace a hyphen with an
+       underscore and a space with a hyphen (and not an underscore).
     """
+    if not key_modifiers:
+        return dict(config)
+
+    # To achieve the replacement strategy, we need a two step process (or
+    # regular expressions):
+    #   1. Replace all occurrencies of the replacement keys with a special
+    #      character (which must not occur in the keys of `config`).
+    #   2. Replace all special characters with the final replacement values.
+    # Note:
+    #   This relies on ordered dictionaries, so Python>=3.6 is required!
+
+    # Make a list of special characters not occurring in `config.keys()`
+    keychars = set()
```

Comment:
I had this idea as well, but what if any developer decides to use `'${0}'` or some other weird characters. I believe, that my approach covers also these border cases – and we only came here, because of unlikely border cases, haven't we? I would have been happy, to just deal with hyphens, spaces, and maybe periods …

Reply:
I don't believe that this is really about border cases. I would argue that this is simply about defining a solid interface. If we would have had only hyphens, spaces and periods, we would also have to decide in which order they are executed or make sure that the replacements do not overlap.

If we use control characters, the probability that these appear in keys should be (close to) zero. Moreover, it can easily be checked for. Maybe I can give this spirit of implementation a try at some point.

Reply:
> If we use control characters, the probability that these appear in keys should be (close to) zero.

But there is still a probability, which is avoided by my approach.

>  Moreover, it can easily be checked for.

That is, what my code does: It creates the control character(s) automatically and checks, if it exists. If it does, it looks for another one. IMHO I believe it's an elegant solution.

It could always find automatically two control characters and but an ever increasing number between them, e.g. we would then have `'\a0\b'`, `'\a1\b'`, if `'\a'` and `'\b'` become suitable control characters. Would that make the code clearer in your view? This can be easily done and reduces the (close to zero) risk of running out of characters even further. :upside_down_face:

Reply:
I went ahead and implemented something by myself. I personally find it a bit more elegant, but I can imagine you don't like it that much.

Since it passes all tests (and therefore arguably implements the desired feature), we could leave it at that.
However, if you have any opinion on (aspects of) my implementation, we can discuss it here.

PS: I hope you don't perceive this as me not being grateful for your contributions. As a matter of fact, my solution is heavily inspired by your work. I just had this feeling it would be a more efficient way of communicating what I had in mind.

Reply:
I am sure a lot of thoughts went into this solution!

You want to hear my honest opinion? It's so complicated, that the functionality does not spring to your eye immediately. (It reminds me of C++ code.) I really do not want to debug it!

I also do not find in-line function definitions (`_chain_replace()`) very elegant. Sorry.

The type of the return value of `_modify_keys()` is still `Mapping` and not `Dict`?

I also do not see the reason behind `(len(_mods) - i)`. Couldn't `(i + 1)` be used to the same results? Also I do not like to have a calculation inside a loop, which does not change its value in its iterations (`len(_mods)`). And finally I was thought by hard experience, that constructs like `_guard + _tmp * (len(_mods) - i) + _guard` have to be avoided at all cost because of penalty in run time (now it's easy to do so by using f-strings).

Finally I do not know, why you do not use `f'{_guard}{i+1}{_guard}'` as intermediate keys? That is, if you really want to keep this approach.

And most important of all: What will happen, if you have `'\x1b\x1a\x1b'` in any of the keys of your `m` mapping (not the `key_mods`)?!

Sorry, but you have asked for my opinion … :innocent:

Maybe a really elegant regular expression is called for?

Reply:
Probably too much thought went into the solution, to the point where I couldn’t see the complexity anymore. 

I would argue that the comparison with LISP would be more appropriate, since I think I used a rather functional approach ;)

The type has been fixed now.

The reason for `len(mod) - i` is that I originally (naively) didn’t use guards. In this case I had to make sure to first replace the longest chains.

The reason I used `_tmp * i` instead of `f”{i}”` is to minimise the nr of possible conflicts.

That is effectively an oversight in the implementation, but should raise an error as well.

Thank you for your opinion. Otherwise I would only have my own…

I am not sure how you would solve this with a regex. However, I’m open for suggestions.